torch/fx/passes/utils/fuser_utils.py

```diff
@@ -220,22 +220,34 @@ def insert_subgm(
     submodule_name = sub_gm.__class__.__name__
     gm.add_submodule(submodule_name, sub_gm)
 
-    # Create a call_module node in main graph.
-    module_node = gm.graph.call_module(submodule_name, args=orig_inputs, kwargs=None)
+    def last_node(target_nodes: tuple[Node, ...]) -> Node:
+        for node in reversed(gm.graph.nodes):
+            if node in target_nodes:
+                return node
 
-    output_node = sub_gm.graph.output_node()
-    if len(orig_outputs) == 1 and not isinstance(output_node.args[0], tuple):
-        # main_remapping[comp.orig_outputs[0]] = module_node
-        orig_outputs[0].replace_all_uses_with(module_node, propagate_meta=True)
-    else:
-        for i, orig_output in enumerate(orig_outputs):
-            # Use Proxy to record getitem access.
-            proxy_out = torch.fx.Proxy(module_node)[i].node  # type: ignore[index]
-            orig_output.replace_all_uses_with(proxy_out, propagate_meta=True)
+    last_input_node: Node = last_node(orig_inputs)
 
-        module_node.meta["val"] = tuple(
-            orig_output.meta.get("val", None) for orig_output in orig_outputs
+    # Create a call_module node in main graph.
+    with gm.graph.inserting_after(last_input_node):
+        module_node = gm.graph.call_module(
+            submodule_name, args=orig_inputs, kwargs=None
         )
+        output_node = sub_gm.graph.output_node()
+
+    next_node = module_node.next
+    with gm.graph.inserting_before(next_node):
```

Comment:
OK, I sat down and understood the algorithm. The reason I think this doesn't work is because you can be in this situation:

```
input0 = ..
input1 = ...
output0 = ...
not_in_subgraph = f(output0)
input2 = ...
output1 = ...
```

Here, we DO need to do some code motion for the not_in_subgraph node, since it is "inside" the set of nodes we want to fuse and we need to push it to the end. But it's not necessarily obvious if we should push an arbitrary node in the middle of the fusion region to the end or the beginning. If instead it didn't depend on output0 but produced input2, then we would need to move it to the beginning. But I think you can make this work.

Reply:
indeed, the UT failure caught this case.

I changed the impl to use stable topo sort instead.